Add  a developer guide document (#42)
cps-org/cps-config: docs/dev.md

For example:

```
meson: bumpt gtest requirement from 1.12 to 1.15

gtest 1.13 fixes a bug on Windows that is affecting our CI, see
google/googletest#1234.

gtest 1.14 adds a new feature that allows for faster test execution that will be
used in the next patch
```

Finally, if a commit fixes any github issues, they should add that,

```
Fixes: #7
```

### Merging and rebasing

Rebasing should generally be preferred to merges. Rebases create a linear
history that is easy to bisect over, and easy for a human to understand. In a
project like cps-config that lacks large subsystems this is easy to maintain.

Merges may be appropriate for large additions or refactors that have moved
quickly and may not have a clean history. This should generally be unnecessary.
Merges also do not go through CI, and as such it's possible to create a state
where main before the merge builds successfully, and the topic branch before the
merge builds successfully, but the merge commit itself is broken.

Back merges (merging the main branch back into a topic branch) should never
happen. This makes bisects particularly hard and the history very messy. A
rebase should be done in this case

## Editors

> Use [editorconfig](https://editorconfig.org). Many editors support editorconfig

Reviewer: I'm using sublime text and this setting from your file:

```
insert_final_newline = false
```

is causing sublime text to strip final newlines from files. You can see the result of that in the PR I submitted.

> natively, others have plugins. Otherwise, configure your editor to use the same

Reviewer 2: ```suggestion
natively. Others have plugins. Otherwise, configure your editor to use the same
```

settings.

## All files

> All files should have an [SPDX license

Reviewer 3: It would be helpful to qualify what files would need a SPDX license header and what others wouldn't, unless the intention here is to say literally all files. In that case we should just fix the existing ones that don't have this header, e.g. `meson.build`, all the cps files used for unit tests, etc.

header](https://spdx.dev/learn/handling-license-info/), along with the relevant
copyright holder information, as long as they have a comment syntax. JSON files
do not, and as such should not have such a header.

For example, in a C or C++ file:
```cpp
// SPDX-License-Identifier: MIT
// Copyright 2024 Copyright Holder
```

## Meson

Use the following naming conventions:

- external_dependencies (found by `dependency()` or `compiler.find_library()`)
  should be prefixed with `dep_`, ex: `dep_fmt`

> - internal dependencies (`declare_dependency()`) should be prefixed with `idep_`

Reviewer 2: Just to catch me up... why? In my mind it's cleaner for each unit of code, even subdirectories, to just treat dependencies like dependencies.

Author: I prefer to use the `idep_` identifier in part for historical reasons. But probably doesn't make sense anymore, so I'll change that.

If you're into history, read on:

In ye old days, a consumer of a subproject would make as series of calls like this, pull a named variable from a subproject into their project:
```meson
sub_gtest = subproject('gtest')
dep_gtest = sub_gtest.get_variable('dep_gtest')
```

In the modern approach, a producer does something like:
```meson
dep_gtest = declare_dependency(...)
meosn.override_dependency('gtest', dep_gtest)
```
and then the consumer just does:
```meson
dep_gtest = dependency('gtest')
```
and Meson will do the right thing.

- external programs (`find_program()`) should be prefixed with `prog_`

> use spaces around `:`, ex: `dependency('gtest', version : '>= 1.5')`

Reviewer 2: ```suggestion
Use spaces around `:`. For example: `dependency('gtest', version : '>= 1.5')`
```

> always drop arguments to a new line when they do not fit on one:

Reviewer 2: ```suggestion
Always put arguments on a new line when they do not fit on one line:
```

```meson
project(
  'cps-config',
  'cpp',
  version : '0.0.1',
)
```

> ## C and C++

Reviewer 2: I was thinking we'd just pick an existing coding standard? Then we could just say "we use the LLVM coding standard", set up an analyzer, and move on?

Prefer C++ (`//`) comments to C-style (`/* */`) in general.

C style comments may be used in the following situations:
  - spec citations

>   - algorithm descriptions

Reviewer 2: I understand what this text says. I really don't understand the reasoning behind it though.

For example:

```c++
/* Do X before Y, unless Z.
 *
 * Per the CPS Spec (0.9.0):
 *
 *   When doing thing, X must be resolved before Y, exception when Z is true,
 *   then Y must be done and X must not.
 *
 */
void do_thing(bool x, bool y, bool z);
```

> `#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
> `CPS_CONFIG_USE_UNDEFINED`

Reviewer 2: ```suggestion
`#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
`CPS_CONFIG_USE_UNDEFINED`
```

Well, this is subtle. If we're going to ship a `cps` library, the prefix would be `CPS_`. The executable could also use `CPS_*` because it's shipping in the same repo so collisions are unlikely-to-impossible. Or it could use `CPS_CONFIG_` if we want to treat `cps` and `cps-config` as two projects sharing one build system and repo.

I'm flexible, but I did want to point out that `CPS_CONFIG_FOOBAR` would be odd in a `cps/foobar.hpp`.

## Tests

> Complex implementation details may use compiled unit tests, like gtests. Which
> are generally faster and easier to debug than functional tests.

Reviewer 2: ```suggestion
Complex implementation details may use compiled unit tests, like gtests. Unit tests
are generally faster and easier to debug than functional tests.
```

> Functional tests should use the tests/runner.py, by adding test cases to the

Reviewer 3: Off topic, and I'm happy to open a separate issue if this calls for its own discussion. Apart from unit and function tests, would it add value to this project to add some more integrative tests, e.g. can we build an interesting project with `cps-config` and some hand-crafted cps files

> Functional tests should use the tests/runner.py, by adding test cases to the
> `tests/cases.toml` along with adding or modifying the provided `.cps` files.

Reviewer 2: ```suggestion
Functional tests should use `tests/runner.py` by adding test cases to
`tests/cases.toml`  along with changes to the accompanying `.cps` files.
```
